### DISCUSSION TITLE
Confusion about NeptuneLogger.save_dir implementation

### DISCUSSION THREAD

**Discussion by daniellepintz** [Lightning-AI/pytorch-lightning#11766] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
In the docstring it says the save_dir is None, but then why does it return a path? Should we change either the docstring, or the implementation here?

https://github.com/PyTorchLightning/pytorch-lightning/blob/9d8faecdb2b873b52b95f2772f4bf48068a0af9a/pytorch_lightning/loggers/neptune.py#L516-L524

**Comment by daniellepintz** [comment]
Any insights on this @awaelchli @edward-io @Borda @rohitgr7?

**Reply by rohitgr7** [reply]
I haven't tried Neptune. Can you try running it in online and offline modes and see where it saves the logs. I also remember Neptune doesn't allow configuring the save_dir. If it saves inside `.neptune` all the time, then docstrings can be updated.

**Reply by daniellepintz** [reply]
My understanding from reading the code is that Neptune doesn't have an offline mode: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/neptune.py

**Comment by awaelchli** [comment]
It would be best to ask the neptune team directly @kamil-kaczmarek @Raalsky @Blaizzy
Neptune doesn't log anything to a local directory so save_dir there doesn't make much sense, so I guess at the time of implementation, the best other option was to point at the ".neptune" folder.

**Reply by daniellepintz** [reply]
It seems the save_dir implementation used to return None before https://github.com/PyTorchLightning/pytorch-lightning/pull/6867. Looks like the PR author forgot to update the docstring. But still a bit confused because I thought Neptune doesn't save anything locally.

**Reply by Blaizzy** [reply]
Hi @daniellepintz 

Prince Canuma here, a Data Scientist at Neptune.ai

I will let the engineering team know about this,

By default, Neptune will create a `'.neptune'` folder inside the current working directory. In the case of #6867 it changes the model checkpoint path to be `'.neptune'` folder in case the user doesn't define his own path using ModelCheckpointCallback() for example.
Check this commit: https://github.com/PyTorchLightning/pytorch-lightning/pull/6867/commits/5ac80ece048c16f19e9c05b1873a08fe0a37be90

>But still a bit confused because I thought Neptune doesn't save anything locally.

Neptune uses the `'.neptune'` folder to store metadata temporarily. For example, you track a run in **offline mode** or there is a network connectivity issue in which case neptune also automatically switches to **offline mode** and save the data to disk. Later you can synchronize the locally stored metadata with the servers using the `neptune sync` CLI command.  

Docs:
- https://docs.neptune.ai/api-reference/command-line-interface#neptune
- https://docs.neptune.ai/api-reference/command-line-interface#neptune-sync

**Reply by daniellepintz** [reply]
thanks so much for the explanation!

**Reply by Blaizzy** [reply]
Most welcome! 

Feel free to reach out if you have any further questions. I would be happy to help👍🏽

**Reply by Blaizzy** [reply]
Hi @daniellepintz 

I hope you are well,

If the answer I gave you satisfies you, do you mind marking it as the answer so others in the community with the same doubt can easily find it? 👍 😃

**Reply by daniellepintz** [reply]
of course!

**Reply by daniellepintz** [reply]
Actually it doesn't seem that is possible since it is not top-level. Feel free to make a new top level comment and I will do so

**Reply by Blaizzy** [reply]
Thank you, I didn't know that.
No worries, a new top-level comment is done ✅ 😊 
https://github.com/PyTorchLightning/pytorch-lightning/discussions/11766#discussioncomment-2274385

**Comment by Blaizzy** [comment, marked as answer]
Hi @daniellepintz

Prince Canuma here, a Data Scientist at Neptune.ai

I will let the engineering team know about this,

By default, Neptune will create a '.neptune' folder inside the current working directory. In the case of https://github.com/PyTorchLightning/pytorch-lightning/pull/6867 it changes the model checkpoint path to be '.neptune' folder in case the user doesn't define his own path using ModelCheckpointCallback() for example.
Check this commit: https://github.com/PyTorchLightning/pytorch-lightning/commit/5ac80ece048c16f19e9c05b1873a08fe0a37be90

> But still, a bit confused because I thought Neptune doesn't save anything locally.

Neptune uses the '.neptune' folder to store metadata temporarily. For example, you track a run in offline mode or there is a network connectivity issue in which case neptune also automatically switches to offline mode and save the data to disk. Later you can synchronize the locally stored metadata with the servers using the neptune sync CLI command.

Docs:

https://docs.neptune.ai/api-reference/command-line-interface#neptune
https://docs.neptune.ai/api-reference/command-line-interface#neptune-sync
